cuda/solver/cb_gmres_kernels.cu

```diff
@@ -0,0 +1,534 @@
+            iter + 1, dim_size[0], dim_size[1],
+            as_cuda_type(next_krylov_basis->get_const_values()),
+            stride_next_krylov, as_cuda_accessor(krylov_bases),
+            as_cuda_type(hessenberg_iter->get_values()), stride_hessenberg,
+            as_cuda_type(stop_status));
+    } else {
+        singledot_kernel<singledot_block_size>
+            <<<grid_size_iters_single, block_size_iters_single>>>(
+                iter + 1, dim_size[0],
+                as_cuda_type(next_krylov_basis->get_const_values()),
+                stride_next_krylov, as_cuda_accessor(krylov_bases),
+                as_cuda_type(hessenberg_iter->get_values()), stride_hessenberg,
+                as_cuda_type(stop_status));
+    }
+    // for i in 1:iter
+    //     hessenberg(iter, i) = next_krylov_basis' * krylov_bases(:, i)
+    // end
+    update_next_krylov_kernel<default_block_size>
+        <<<ceildiv(dim_size[0] * stride_next_krylov, default_block_size),
+           default_block_size>>>(
+            iter + 1, dim_size[0], dim_size[1],
+            as_cuda_type(next_krylov_basis->get_values()), stride_next_krylov,
+            as_cuda_accessor(krylov_bases),
+            as_cuda_type(hessenberg_iter->get_const_values()),
+            stride_hessenberg, as_cuda_type(stop_status));
+
+    // for i in 1:iter
+    //     next_krylov_basis  -= hessenberg(iter, i) * krylov_bases(:, i)
+    // end
+    components::fill_array(exec, arnoldi_norm->get_values() + stride_arnoldi,
+                           dim_size[1], zero<non_complex>());
+    if (use_scalar) {
+        components::fill_array(exec,
+                               arnoldi_norm->get_values() + 2 * stride_arnoldi,
+                               dim_size[1], zero<non_complex>());
+    }
+    multinorm2_inf_kernel<use_scalar><<<grid_size, block_size>>>(
+        dim_size[0], dim_size[1],
+        as_cuda_type(next_krylov_basis->get_const_values()), stride_next_krylov,
+        as_cuda_type(arnoldi_norm->get_values() + stride_arnoldi),
+        as_cuda_type(arnoldi_norm->get_values() + 2 * stride_arnoldi),
+        as_cuda_type(stop_status));
+    // nrmN = norm(next_krylov_basis)
+    components::fill_array(exec, num_reorth->get_data(), 1, zero<size_type>());
+    check_arnoldi_norms<default_block_size>
+        <<<ceildiv(dim_size[1], default_block_size), default_block_size>>>(
+            as_cuda_type(arnoldi_norm->get_values()), stride_arnoldi,
+            as_cuda_type(hessenberg_iter->get_values()), stride_hessenberg,
+            iter + 1, as_cuda_accessor(krylov_bases), as_cuda_type(stop_status),
+            as_cuda_type(reorth_status), as_cuda_type(num_reorth->get_data()));
+    numReorth = 0;
+    exec->get_master()->copy_from(exec.get(), 1, num_reorth->get_const_data(),
+                                  &numReorth);
+    // numReorth <= number of next_krylov vector to be reorthogonalization
+    for (size_type l = 1; (numReorth > 0) && (l < 3); l++) {
```

Comment:
I don't really know why the `3` was added, but so far, but we do only perform the re-orthogonalization when necessary. I would prefer to keep it as is, and only change it when it is clear that this is a source for a problem.

Reply:
From @josealiaga
> As Pratik Nayak says, the number 3 is the typical value in the literature ([link](https://www.sciencedirect.com/science/article/pii/S0898122105003366)). It could be defined as a parameter, but I don't know if it makes sense ...